.github/actions/run_awx_devel/action.yml

```diff
@@ -0,0 +1,92 @@
+# This currently *always* uses the "warm build cache" image
+# We should do something to allow forcing a rebuild, probably by looking for
+# some string in the commit message or something.
+
+name: Run AWX (devel environment)
+description: Runs AWX with `make docker-compose`
+inputs:
+  github-token:
+    description: GitHub Token for registry access
+    required: true
+  build-ui:
+    description: Should the UI be built?
+    required: false
+    default: false
+    type: boolean
+outputs:
+  ip:
+    description: The IP of the tools_awx_1 container
+    value: ${{ steps.data.outputs.ip }}
+  admin-token:
+    description: OAuth token for admin user
+    value: ${{ steps.data.outputs.admin_token }}
+runs:
+  using: composite
+  steps:
+    - name: Get python version from Makefile
+      shell: bash
+      run: echo py_version=`make PYTHON_VERSION` >> $GITHUB_ENV
+
+    - name: Remove built-in ansible-core
+      shell: bash
+      run: /usr/bin/python3 -m pip uninstall -y ansible-core
+
+    - name: Install python ${{ env.py_version }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ env.py_version }}
+
+    - name: Install ansible-core with new python
+      shell: bash
+      run: pip3 install ansible-core
+
+    - name: Install system deps
+      shell: bash
+      run: sudo apt-get install -y gettext
+
+    - name: Log in to registry
+      shell: bash
+      run: |
+        echo "${{ inputs.github-token }}" | docker login ghcr.io -u ${{ github.actor }} --password-stdin
+
+    - name: Pre-pull image to warm build cache
```

Comment:
It looks like you're just running out of that image, not just warming up your cache.

Reply:
That comment/name came from the original e2e_tests, IIRC. But yeah, valid, I can change it.

Actually, I think it'd be really cool to cache the *database* every once in a while. Then just let things fetch a relatively recent database and apply the latest few migrations on top of it.

But that's a rainy-day idea.

Reply:
So I've updated it so that it builds the image each time now from the current checkout.

I guess we do want that (consider: developing a feature that also has collection changes, we want to test them in lockstep). But since it adds so much time, I might split up the job matrix a bit more to try to get the time back down.

Reply:
Thinking more on this and doing some more research -- we're using buildkit's inline caching which only caches the last stage and we use a multistage build. So I think we don't really get to take advantage of the caching here.

I'd like to propose (or at least explore and discuss) making our "official" ghcr.io devel images (the ones that get built on pushes to devel and feature branches) use a [registry cache](https://docs.docker.com/build/cache/backends/registry/) instead. Basically they would push a secondary cache image alongside the branch image.

So we'd have (for example): `ghcr.io/ansible/awx_devel:devel` and `ghcr.io/ansible/awx_devel:devel_cache` or something.

I can't guarantee this will speed things up, but I think there's a chance that it could.

Reply:
At first glance, I don't think I disagree with your proposal for use of the registry cache, but we certainly want to keep that out of the current scope and I'd like @shanemcd to have a look. What we already have has been working _relatively_ well to keep build times fast and there are sometimes slowdown you see due to out-of-date branches or dependency changes that don't apply to small patches and aren't a real problem.